backend/witness/vector.go

```diff
@@ -1,92 +1,288 @@
 package witness
 
 import (
-	"io"
+	"errors"
 	"math/big"
 	"reflect"
 
 	"github.com/consensys/gnark-crypto/ecc"
-	"github.com/consensys/gnark/frontend/schema"
-	witness_bls12377 "github.com/consensys/gnark/internal/backend/bls12-377/witness"
-	witness_bls12381 "github.com/consensys/gnark/internal/backend/bls12-381/witness"
-	witness_bls24315 "github.com/consensys/gnark/internal/backend/bls24-315/witness"
-	witness_bls24317 "github.com/consensys/gnark/internal/backend/bls24-317/witness"
-	witness_bn254 "github.com/consensys/gnark/internal/backend/bn254/witness"
-	witness_bw6633 "github.com/consensys/gnark/internal/backend/bw6-633/witness"
-	witness_bw6761 "github.com/consensys/gnark/internal/backend/bw6-761/witness"
+	fr_bls12377 "github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	fr_bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
+	fr_bls24315 "github.com/consensys/gnark-crypto/ecc/bls24-315/fr"
+	fr_bls24317 "github.com/consensys/gnark-crypto/ecc/bls24-317/fr"
+	fr_bn254 "github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	fr_bw6633 "github.com/consensys/gnark-crypto/ecc/bw6-633/fr"
+	fr_bw6761 "github.com/consensys/gnark-crypto/ecc/bw6-761/fr"
 	"github.com/consensys/gnark/internal/tinyfield"
-	witness_tinyfield "github.com/consensys/gnark/internal/tinyfield/witness"
 	"github.com/consensys/gnark/internal/utils"
 )
 
-type Vector interface {
-	io.WriterTo
-	io.ReaderFrom
-	FromAssignment(assignment interface{}, leafType reflect.Type, publicOnly bool) (*schema.Schema, error)
-	ToAssignment(assigment interface{}, leafType reflect.Type, publicOnly bool)
-	Len() int
-	Type() reflect.Type
-}
-
-func newVector(field *big.Int) (Vector, error) {
-	var w Vector
+func newVector(field *big.Int, size int) (any, error) {
```

Comment:
For understanding the design - in witness we use a pointer to the fr.Vector (which is a slice). Imo slice is struct{head *ptr, cap, len int} (https://pkg.go.dev/reflect#SliceHeader) and I think it isn't a big difference if we work with `*Vector` or `Vector`.

Reply:
yes may be a clear way to avoid it now; at some point in the refactoring it was just easier to do this for deserialization and growing the slice in-place.